python/check_data.py

```diff
@@ -0,0 +1,174 @@
+    Literal["grade"],
+    Literal["test"],
+    Literal["file"],
+]
+
+all_phases: List[Phase] = [
+    "generate",
+    "prepare",
+    "render",
+    "parse",
+    "grade",
+    "test",
+    "file",
+]
+
+
+class PropInfo(TypedDict):
+    type: str
+    present_phases: List[Phase]
+    edit_phases: List[Phase]
+
+
+PROPS: Dict[str, PropInfo] = {
+    "params": {
+        "type": "object",
+        "present_phases": all_phases,
+        "edit_phases": ["generate", "prepare", "grade"],
+    },
+    "correct_answers": {
+        "type": "object",
+        "present_phases": all_phases,
+        "edit_phases": ["generate", "prepare", "parse", "grade"],
+    },
+    "variant_seed": {
+        "type": "integer",
+        "present_phases": all_phases,
+        "edit_phases": [],
+    },
+    "options": {
+        "type": "object",
+        "present_phases": all_phases,
+        "edit_phases": [],
+    },
+    "submitted_answers": {
+        "type": "object",
+        "present_phases": ["render", "parse", "grade"],
+        "edit_phases": ["parse", "grade"],
+    },
+    "format_errors": {
+        "type": "object",
+        "present_phases": ["render", "parse", "grade", "test"],
+        "edit_phases": ["parse", "grade", "test"],
+    },
+    "raw_submitted_answers": {
+        "type": "object",
+        "present_phases": ["render", "parse", "grade", "test"],
+        "edit_phases": ["test"],
+    },
+    "partial_scores": {
+        "type": "object",
+        "present_phases": ["render", "grade", "test"],
+        "edit_phases": ["grade", "test"],
+    },
+    "score": {
+        "type": "number",
+        "present_phases": ["render", "grade", "test"],
+        "edit_phases": ["grade", "test"],
+    },
+    "feedback": {
+        "type": "object",
+        "present_phases": ["render", "grade", "test"],
+        "edit_phases": ["grade", "test"],
+    },
+    "editable": {
+        "type": "boolean",
+        "present_phases": ["render"],
+        "edit_phases": [],
+    },
+    "manual_grading": {
+        "type": "boolean",
+        "present_phases": ["render"],
+        "edit_phases": [],
+    },
+    "panel": {
+        "type": "string",
+        "present_phases": ["render"],
+        "edit_phases": [],
+    },
+    "num_valid_submissions": {
+        "type": "integer",
+        "present_phases": ["render"],
+        "edit_phases": [],
+    },
+    "gradable": {
+        "type": "boolean",
+        "present_phases": ["parse", "grade", "test"],
+        "edit_phases": [],
+    },
+    "filename": {
+        "type": "string",
+        "present_phases": ["file"],
+        "edit_phases": [],
+    },
+    "test": {
+        "type": "string",
+        "present_phases": ["test"],
+        "edit_phases": [],
+    },
+    "extensions": {
+        "type": "object",
+        "present_phases": all_phases,
+        "edit_phases": [],
+    },
+}
+
+
+def check_prop(
+    prop: str,
+    old_value: Any,
+    new_value: Any,
+    value_type: str,
```

Comment:
Should this be a literal as well?

Reply:
It probably can be, this is a well-known set of values. I'll add that.